ddsrecorder_participants/include/ddsrecorder_participants/mcap/McapHandler.hpp

```diff
@@ -0,0 +1,175 @@
+{
+    Message() = default;
+
+    Message(
+            const Message& msg)
+    : mcap::Message(msg)
+    {
+        this->payload_owner = msg.payload_owner;
+        auto payload_owner_ =
+            const_cast<eprosima::fastrtps::rtps::IPayloadPool*>((eprosima::fastrtps::rtps::IPayloadPool*)msg.payload_owner);
+        this->payload_owner->get_payload(
+                msg.payload,
+                payload_owner_,
+                this->payload);
+    }
+
+    ~Message()
+    {
+        // If payload owner exists and payload has size, release it correctly in pool
+        if (payload_owner && payload.length > 0)
+        {
+            payload_owner->release_payload(payload);
+        }
+    }
+
+    ddspipe::core::types::Payload payload{};
+    ddspipe::core::PayloadPool* payload_owner{nullptr};
+};
+
+/**
+ * TODO
+ */
+class McapHandler : public ddspipe::participants::ISchemaHandler
+{
+public:
+
+    McapHandler(
+            const char* file_name,
+            std::shared_ptr<ddspipe::core::PayloadPool> payload_pool,
+            unsigned int max_pending_samples,
+            unsigned int buffer_size,
+            unsigned int downsampling,
+            unsigned int event_window);
+            // bool autostart = false);
+
+    ~McapHandler();
+
+    void add_schema(
+            const fastrtps::types::DynamicType_ptr& dynamic_type) override;
+
+    void add_data(
+            const ddspipe::core::types::DdsTopic& topic,
+            ddspipe::core::types::RtpsPayloadData& data) override;
+
+    void start();
+
+    void pause();
+
+    void trigger_event();
+
+protected:
+
+    // enum StateCode
+    // {
+    //     UNSTARTED = 0,
+    //     STARTED,
+    //     PAUSED,
+    // };
+
+    void add_data_(
+            const Message& msg);
+
+    void add_pending_samples_(
+            const std::string& schema_name);
+
+    void event_thread_routine_();
+
+    void stop_event_thread_();
+
+    void clear_all_();
+
+    void dump_data_();
+
+    void dump_data_nts_();
+
+    mcap::ChannelId create_channel_id_nts_(
+            const ddspipe::core::types::DdsTopic& topic);
+
+    mcap::ChannelId get_channel_id_(
+            const ddspipe::core::types::DdsTopic& topic);
+
+    mcap::SchemaId get_schema_id_(
+            const std::string& schema_name);
+
+    mcap::Timestamp now();
+    mcap::Timestamp fastdds_timestamp_to_mcap_timestamp(const ddspipe::core::types::DataTime& time);
+
+    // StateCode state_ = {UNSTARTED};
+
+    // NOTE: it cannot be used with Atomicable because McapWriter is a final class
+    mcap::McapWriter mcap_writer_;
+    std::mutex write_mtx_;
+
+    using SchemaMapType = utils::SharedAtomicable<std::map<std::string, mcap::Schema>>;
+    SchemaMapType schemas_;
+
+    using ChannelMapType = utils::SharedAtomicable<std::map<std::string, mcap::Channel>>;
+    ChannelMapType channels_;
+
+    using PendingSamplesMapType = utils::SharedAtomicable<std::map<std::string, std::queue<std::pair<std::string, Message>>>>;
```

Comment:
Atomicable is useful and allows to keep joint the variable and its mutex.
However, when there are multiple variables to keep safe, it is more efficient to have a single mutex (mutex block costs, but taken and freeing a mutex is also a performance issue)

Reply:
Simplified to use only one object mutex plus one for concurrency with event routine thread.